### PR TITLE
Compose: Introduce in-house `compose` and `pipe` utils

### PR DESCRIPTION
**Pull request by tyxla**
## What?
Currently, we expose our own `compose()` high-order composition utility, which is a re-export of `_.flowRight()`. This PR implements our simple in-house version of `compose()`, and uses the opportunity to also introduce an in-house version of `flow()`, which is the inverted version of `flowRight()`.  That allows us to get rid of the corresponding Lodash functions in the following PRs.

## Why?
Lodash is known to unnecessarily inflate the bundle size of packages, and in most cases, it can be replaced with native language functionality. See these for more information and rationale:

* https://github.com/WordPress/gutenberg/issues/16938#issuecomment-602837246
* https://github.com/WordPress/gutenberg/issues/17025
* https://github.com/WordPress/gutenberg/issues/39495 

## How?
We're implementing a custom base `baseFlow()` function and using that to generate both `compose()` and `flow()` so they work the same way as the Lodash versions. We're adding a few tests to ensure things work as expected.

I've chosen to migrate away from the existing `_.flowRight()` and `_.flow()` in a separate PR, to keep this one more contained.

## Testing Instructions
* Smoke test all editors - `compose()` is used all over the place.
* Verify all tests pass.

### CODE REVIEW

**Review by chad1008 — COMMENTED** [review]
Changes look good to me, and test fine. I'll defer to more experienced eyes for an approval though 👍

I am curious about the naming. These are inspired by `lodash`'s `flow` and `flowRight`, but we're calling them `flow` and `compose`, which kind of hides that they do the same thing in opposite directions.

Is that just because `compose()` was named and used all over the place already, and renaming all of those implementations to `flowRight()` would be more trouble than it's worth?

**Comment by gziolo** [comment]
> I am curious about the naming. These are inspired by `lodash`'s `flow` and `flowRight`, but we're calling them `flow` and `compose`, which kind of hides that they do the same thing in opposite directions.
> 
> Is that just because `compose()` was named and used all over the place already, and renaming all of those implementations to `flowRight()` would be more trouble than it's worth?

The naming in `lodash` is strange. It’s usually `compose` + `pipe` pair that you will se in the learning materials about functional programming. In fact, [`ramda`](https://ramdajs.com) ("A practical functional library for JavaScript programmers.") uses `compose` and `pipe` in their API.

File: packages/compose/src/higher-order/flow.ts
Comment (gziolo) [inline packages/compose/src/higher-order/flow.ts]: It would be better to include the license from `lodash` like here:

https://github.com/WordPress/gutenberg/blob/a3c753d28ddb0972311e2da0136277626bdf832e/packages/element/src/serialize.js#L1-L26

**Comment by gziolo** [comment]
From my perspective the refactor for `compose` is good to go after we include the license from `lodash`.

As for `flow`, it isn't technically necessary, but I don't mind if we add it as it just reversed the order of execution. I would prefer we name it `pipe`. By the way, there is also a [proposal to include the pipeline operator in JavaScript](https://github.com/tc39/proposal-pipeline-operator), but it's still stage 2.

**Comment by tyxla** [comment]
> I am curious about the naming. These are inspired by `lodash`'s `flow` and `flowRight`, but we're calling them `flow` and `compose`, which kind of hides that they do the same thing in opposite directions.

The original goal behind naming here was to:
* Keep `compose` because that's how it was already named.
* Keep using `flow` as that's how it was referenced originally in existing code (`flow` is still used in a few places).

> Is that just because `compose()` was named and used all over the place already, and renaming all of those implementations to `flowRight()` would be more trouble than it's worth?

Essentially, yes.

File: packages/compose/src/higher-order/flow.ts
Comment (tyxla) [reply packages/compose/src/higher-order/flow.ts]: Done in https://github.com/WordPress/gutenberg/pull/44112/commits/26a930481133a87623cdc64bf9cab3746dd31311

**Comment by tyxla** [comment]
Thanks for the feedback, both 🙌 

As suggested by @gziolo I've renamed `flow` to `pipe` and added the Lodash license. We'll update existing `flowRight` and `flow` usages to use `compose` and `pipe` in subsequent PRs.

As for the native JS pipeline operator proposal - I was also considering using it (and making sure Babel supports it), but I saw there's still no consensus on the syntax. Maybe one day we can update to it, when it finally lands in the ES world.

That being said, this PR is ready for another final look.

**Review by gziolo — APPROVED** [review]
Looks great. I'm happy to see that included in `@wordpress/compose`. Excellent work @tyxla 💯

**Comment by tyxla** [comment]
I realized I needed to add a CHANGELOG entry. Added in https://github.com/WordPress/gutenberg/pull/44112/commits/9aa5a36b0c5b7d31afc1e58efa4bab65f0e94e19

**Pull request merged**
